### ISSUE TITLE
Unable to Import the module in Azure Automation Runbook

### ISSUE THREAD

**Issue by octokarol** [EvotecIT/Mailozaurr#58]
Hey,
I'm unable to import this module in Az Automation, both in 7.2 runtime and 5.1 runtime. Tested on a "clean" Runtime Environment without any other modules installed. The error i get is:

Import-Module: Line | 2 | Import-Module Mailozaurr | ~~~~~~~~~~~~~~~~~~~~~~~~ | Could not load file or assembly 'System.CodeDom, Version=7.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'.

Is this a limitation of Az Automation?

**Comment by PrzemyslawKlys** [comment]
Try using older version, i think System.CodeDom was added in last version.

**Comment by PrzemyslawKlys** [comment]
![image](https://github.com/user-attachments/assets/e40eefa2-af46-441b-aa7f-3bab85ca4d97)

I need to investigate why this library is there and whether this is actually needed still

**Comment by octokarol** [comment]
thanks
tested 1.0.2 and it works fine
i've also checked the 2.0.0 release from ps gallery and it also didn't work

**Comment by PrzemyslawKlys** [comment]
it has probably the same dependency, i need to investigate whether this library is even necessary or could be exclude

**Comment by PrzemyslawKlys** [comment]
can you try new version and see if it works?

**Comment by octokarol** [comment]
1.0.4 does not work
Import-Module: Could not load file or assembly 'System.Formats.Asn1, Version=8.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'.

**Comment by PrzemyslawKlys** [comment]
![image](https://github.com/user-attachments/assets/e1472e34-7db2-4e66-820e-a156daa3dc07)

Ok, this one is directly related to Mailkit/miimekit so may be a bit harder to "remove". And the question is why it doesn't load

**Comment by PrzemyslawKlys** [comment]
@jstedfast Do you maybe know if that System.Formats.Asn1 is critical? Or I could potentially exclude it like I did for above?

https://github.com/EvotecIT/Mailozaurr/blob/6cd7430f0dde6ca428792dabc389963e8796d210/Sources/Mailozaurr.PowerShell/Mailozaurr.PowerShell.csproj#L18-L22

I am not sure why would Azure.Automations complain about it tho

**Comment by PrzemyslawKlys** [comment]
Looking into it seems to be related to certs so probably can't do much about it:


> Provides classes that can read and write the ASN.1 BER, CER, and DER data formats.
> 
> Commonly Used Types:
> System.Formats.Asn1.AsnReader
> System.Formats.Asn1.AsnWriter

More important part is why would it complain on Azure.Automations

**Comment by octokarol** [comment]
was this package (Asn1.AsnWriter) also included in 1.0.2?
for my use case I can keep using this version basically indefinitely, all I use it for is a wrapper for Send-MgUserMail so I can keep my sanity (i actually don't know if it uses send-mgusermail under the hood for Graph mailing, just assumed so)

**Comment by octokarol** [comment]
let me know if there's any way i could collect additional logs from azure automations runtime environment that could help

**Comment by PrzemyslawKlys** [comment]
No, it's only added in 1.0.3+

![image](https://github.com/user-attachments/assets/cf97d308-c333-452e-a532-1fc3cef0882f)

I've changed how I build the module by basically creating a project in C# and telling it to publish so it gathers all relevant DLL's and it's much easier to find what works, what doesn't. Up to 1.0.2 I was basically doing it manually which works, but at the same time is pain to deal with what's missing. 

I guess I could remove Asn1 and it would probably be fine, but maybe some feature would crash the library later on. I could also just skip it from loading and potentially it would load itself when necessary. I may need to work with it somehow. 

It doesn't use Send-mgusermail, it uses Invoke-MgRestMethod if you're using MgGraph, and if not it uses Invoke-RestMethod, as it's doing fairly bit more than what that cmdlet offers. In your case Mailozaurr doesn't even use Mailkit/MimeKit so DLLs are not really needed, but I'll try to solve it anyway

**Comment by octokarol** [comment]
perhaps it could be only loaded when not running in azure automation?
https://stackoverflow.com/questions/37224874/what-is-the-correct-way-to-detect-whether-a-script-is-running-in-azure-automatio
it seems that it is possible to detect the runtime enviroment

let me know if there's any additional information i could collect to help you troubleshoot this

**Comment by PrzemyslawKlys** [comment]
```powershell
if ($PSPrivateMetadata.JobId) {
    # Azure Automation
}
else {
    # Local
}
```

Could be a solution, yes - but finally at the end of the day something probably won't run ;) In your case it will work - but who knows why it doesn't work. It ould make sense to understand why it fails there, but works in local environment. Maybe some NET has to be installed

**Comment by PrzemyslawKlys** [comment]
Are you able to load module from .zip or something? or do you need it published? i have 2-3 things to test

**Comment by PrzemyslawKlys** [comment]
- 1st idea - move those entries to the top of the libraries to load and see if that order matters for azure 
![image](https://github.com/user-attachments/assets/b8bc3b47-55a5-4ae1-9874-5accd6fcfe25)
- 2nd idea - I think I am missing one class Initialize that I use in other modules so I can republish this to make sure to use it

**Comment by PrzemyslawKlys** [comment]
Also did it fail on both PS 7 and PS 5.1 or you tested latest on just one? It seems System.Formats.Asn1.dll is part of PowerShell 7, so could be removed, but does it fail in both?

**Comment by octokarol** [comment]
I've tested 1.0.4 just on 7.2, as some other modules I need work better on it. I can upload anything from a .zip for the runtime.

**Comment by PrzemyslawKlys** [comment]
Try 5.1 as it will tell us if it's conflict with PS 7, because it brings it's own version, or it's something else.

**Comment by octokarol** [comment]
I cannot test my full script on 5.1 as Az and Graph modules I use are quite grumpy on it, but it seems that the Import is successful and no errors show up on the output. Tested on a 5.1 env with only Az and Mailozaurr 1.0.4 imported.

**Comment by PrzemyslawKlys** [comment]
ok, that makes sense now. I'll try to release new version, after some testing and we will see

**Comment by PrzemyslawKlys** [comment]
So it seems the newest version that I have seems to work on PS 7.4, but not on PS 7.2. PS 7.2 LTS end support date is in 4 days. Is there no 7.4 on azure automation?
- https://azure.microsoft.com/en-us/updates/public-preview-powershell-74-support-for-azure-functions/

**Comment by octokarol** [comment]
When I created this issue 7.2 was the latest version available, but right now 7.4 seems to be an option too. Must have been added in the last 4 days. I'll check 1.0.4 on 7.4 az automation and report back.

**Comment by PrzemyslawKlys** [comment]
tnx, i will also release new version if that other one won't work that does additional changes

**Comment by octokarol** [comment]
After doing drop-in replacement of 1.0.2 to 1.0.4 and PowerShell 7.2 to 7.4, I get "Could not load file or assembly 'Microsoft.Identity.Client, Version=4.61.3.0, Culture=neutral, PublicKeyToken=0a613f4dd989e8ae". After rolling back to 1.0.2 while keeping PowerShell 7.4 everything is back to working.

 I don't get this error when importing 1.0.4 on a clean env. It might be a conflict with MSAL which I also use. However, 1.0.2 continues to work fine with all my other modules.

**Comment by PrzemyslawKlys** [comment]
So Microsoft.Identity is a problem on it's own. The issue comes from multiple modules and multiple versions as they use it and it's a nightmare to solve:
- Exchange module
- Graph module 
- Pnp PowerShell module
- Msal.PS module
- Mailozaurr
- DbaTools
- Az.Accounts

All those modules and probably few others deliver this.

```powershell
$Folder = $Env:PSModulePath -split ";"
$Output1 = foreach ($F in $Folder) {
    if (Test-Path -Path $F) {
        Get-ChildItem -LiteralPath $F -Recurse -Filter "*.dll" | ForEach-Object {
            if ($_.Name -like "Microsoft.IdentityModel.Abstractions.dll") {
                [PSCustomObject] @{
                    Name    = $_.FullName
                    Version = $_.VersionInfo.FileVersion
                }
            }
        }
    }
}
$Output1 | Format-Table -AutoSize

# Could not load type 'Microsoft.Identity.Client.Extensions.Msal.MsalCachePersistenceException
$Output2 = foreach ($F in $Folder) {
    if (Test-Path -Path $F) {
        Get-ChildItem -LiteralPath $F -Recurse -Filter "*.dll" | ForEach-Object {
            if ($_.Name -like "Microsoft.Identity.Client.*") {
                [PSCustomObject] @{
                    Name    = $_.FullName
                    Version = $_.VersionInfo.FileVersion
                }
            }
        }
    }
}
$Output2 | Format-Table -AutoSize
```

You can run this to check what versions you have and how it looks.

Identity.DLL

![image](https://github.com/user-attachments/assets/e07243c7-32fa-4461-8bbb-a3985039a03f)

Identity.Abstractions.DLL (which will also be a problem):

![image](https://github.com/user-attachments/assets/7d51d8b5-4bfd-4886-8320-0511c65b5862)

Now depending on which version of which modules you use it's going to be a huge bummer. The biggest problem is with MSAL.PS because it actually is deprecated and no longer updated so they don't update DLL's anymore.  

For example newest Pnp.PowerShell (2.12) is already with higher version than AZ.Accounts, ExchangeOnlineManagement or/and Microsoft.Graph.Authentication, so I use the 2.10 myself, at least until other modules cach up. I started dropping MSAL.PS everywhere, or additionally replacing Identity.Client.dll with a modern, up to date version but that usually is a bit more complicated as it has other dependencies so one has to be careful. 

What also worked for me in the past is different order of loading modules and using PS 5.1. But you never know what will work now. It's a nightmare

**Comment by octokarol** [comment]
That's terrible. Thanks for the detailed explanation. I assume that the issue can be closed as it is indeed solved on 7.4 and 1.0.4 - importing Mailozaurr on its own works.
